garage/logger/outputs.py

```diff
@@ -0,0 +1,178 @@
+"""Contains the output classes for the logger.
+
+Each class is sent logger data and handles it itself.
+"""
+import abc
+import csv
+import datetime
+import os
+import sys
+
+import dateutil.tz
+
+from garage.logger.tabular_input import TabularInput
+from garage.misc.console import mkdir_p
+
+
+class LogOutput(abc.ABC):
+    """Abstract class for Logger Outputs."""
+
+    @property
+    def types_accepted(self):
+        """Pass these types to this logger output.
+
+        The types in this tuple will be accepted by this output.
+
+        :return: A tuple containing all valid input types.
+        """
+        return ()
+
+    @abc.abstractmethod
+    def record(self, data, prefix=''):
+        """Pass logger data to this output.
+
+        :param data: The data to be logged by the output.
+        :param prefix: A prefix placed before a log entry in text outputs.
+        """
+        pass
+
+    def dump(self, step=None):
+        """Dump the contents of this output.
+
+        :param step: The current run step.
+        """
+        pass
+
+    def close(self):
+        """Close any files used by the output."""
+        pass
+
+    def __del__(self):
+        """Clean up object upon deletion."""
+        self.close()
+
+
+class NullOutput(LogOutput):
+    """Dummy output to disable 'no logger output' warnings."""
+
+    @property
+    def types_accepted(self):
+        """Accept all output types."""
+        return (object, )
+
+    def record(self, data, prefix=''):
+        """Don't do anything."""
+        pass
+
+
+class StdOutput(LogOutput):
```

Comment:
it looks like this class is completely uncovered during testing?

Reply:
I totally forgot I could just redirect standard out to test this